app/utils/graphql.ts

```diff
@@ -27,40 +29,74 @@ const DEFAULT_SWR_OPTIONS: SWRConfiguration = {
 /**
  * Fetcher which interacts with our GraphQL API
  */
-const fetcher = <T>([query, variables, token, hasuraRoleName]: [
+const fetcher = <T>([query, variables, token, hasuraRoleName, schema]: [
   RequestDocument,
   Variables,
   string,
-  string
+  string,
+  ZodSchema<T>
 ]): Promise<T> =>
-  request({
+  request<T>({
     url: ENDPOINT,
     document: query,
     variables,
     requestHeaders: {
       Authorization: `Bearer ${token}`,
       "x-hasura-role": hasuraRoleName,
     },
+  }).then((data) => {
+    return schema.parseAsync(data);
   });
 
+/**
+ * Hook which wraps our data schema in an outer object that
+ * matches what Hasura will return. e.g., it shapes a Crash
+ * object schema into a shape like
+ * { crashes: <Crash[]> }
+ */
+const useApiResponseSchema = <T extends Record<string, unknown>>(
+  schema: ZodSchema<T>,
+  typename: string
+): ZodSchema<{ [key in typeof typename]: T[] }> =>
+  useMemo(() => {
+    return z.object({ [typename]: z.array(schema) });
+  }, [schema, typename]);
+
+interface UseQueryResponse<T> {
+  data?: T[];
+  error: any;
+  isLoading: boolean;
+  isError: boolean;
+  refetch: () => Promise<any>;
+  isValidating: boolean;
+}
 /**
  * Hook which wraps `useSWR` and provides auth
+ *
+ * This hook is limited to querying just one type per query as
+ * will be defined by the typename
  */
-export const useQuery = <T>({
+export const useQuery = <T extends Record<string, unknown>>({
   query,
   variables,
   options,
+  schema,
+  typename,
 }: {
   query: RequestDocument | null;
   variables?: Variables;
   options?: SWRConfiguration;
-}) => {
+  typename: string;
+  schema: ZodSchema<T>;
+}): UseQueryResponse<T> => {
   const { getIdTokenClaims, user } = useAuth0();
 
+  const responseSchema = useApiResponseSchema(schema, typename);
```

Comment:
this is how we handle the actual shape of the Hasura response, which returns something that looks like 

```json
{
  "crashes_list_view": [
    {
      "record_locator": "20484424",
      "case_id": "243100443",
      "crash_timestamp": "2024-11-05T14:40:00+00:00",
      "address_primary": "4700 N 35 HWY",
      "collsn_desc": "SAME DIRECTION - BOTH GOING STRAIGHT-REAR END"
    }
  ]
}
```